miniapp/miniapp_tsgemm.cpp

```diff
@@ -0,0 +1,484 @@
+                                             : thread_priority::thread_priority_default;
+  ExecutorType mpi_executor(pool_name, priority);
+#endif
+
+  // Benchmark calls of `sirius_gemm`
+  //
+  for (int i = 0; i < ps.num_iters; ++i) {
+    MPI_Barrier(MPI_COMM_WORLD);
+    auto t_start = clock_t::now();
+
+    sirius_gemm(mpi_executor, comm_grid, a_mat, b_mat, cini_mat, cfin_mat);
+    waitall_tiles(cfin_mat);
+
+    MPI_Barrier(MPI_COMM_WORLD);
+    auto t_end = clock_t::now();
+
+    if (rank == 0) {
+      std::printf("%d: t_tot  [s] = %.5f\n", i, seconds_t(t_end - t_start).count());
+    }
+
+    // Simple check
+    if (ps.check) {
+      ScalarType cfin_sum = sum_matrix(comm_world, cfin_mat);
+      if (rank == 0) {
+        std::cout << cfin_sum << '\n';
+      }
+    }
+  }
+
+  // Upon exit the mpi/user polling RAII object will stop polling
+  return hpx::finalize();
+}
+
+// Example usage:
+//
+//   mpirun -np 1 tsgemm --len_m      100  --len_n      100  --len_k  10000
+//                       --tile_m      32  --tile_n      32
+//                       --pgrid_rows   1  --pgrid_cols   1
+//
+int main(int argc, char** argv) {
+  // Flush printf
+  setbuf(stdout, nullptr);
+
+  // Initialize MPI
+  int thd_required = MPI_THREAD_MULTIPLE;
+  int thd_provided;
+  MPI_Init_thread(&argc, &argv, thd_required, &thd_provided);
+
+  if (thd_required != thd_provided) {
+    std::fprintf(stderr, "Provided MPI threading model does not match the required one.\n");
+    MPI_Abort(MPI_COMM_WORLD, 1);
+  }
+
+  // Declare options before creating resource partitioner
+  namespace po = hpx::program_options;
+  po::options_description desc_cmdline = ::init_desc();
+  po::variables_map vm;
+  po::store(po::command_line_parser(argc, argv).allow_unregistered().options(desc_cmdline).run(), vm);
+  std::string setup = vm["setup"].as<std::string>();
+
+  if (setup == "mpi_pool") {
+    int num_procs;
+    MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
+    if (num_procs > 1) {  // if more than 1 process
+      // Create a thread pool for MPI work and add (enabled) PUs on the first core
+      //
+      // Note: Thread pools must be declared before starting the runtime
+      hpx::resource::partitioner rp(desc_cmdline, argc, argv);
+      if (rp.numa_domains()[0].cores().size() != 1) {  // if more than 1 core
+        rp.create_thread_pool("mpi");
+        rp.add_resource(rp.numa_domains()[0].cores()[0].pus(), "mpi");
+      }
+    }
+  }
+
+  // Start the HPX runtime
+  hpx::init(desc_cmdline, argc, argv);
+
+  MPI_Finalize();
+}
+
+namespace {
+
+// Send Cini tile to the process it belongs (`tile_rank`)
+void send_tile(ConstTileType const& c_tile, SizeType tile_rank, SizeType tag, MPI_Comm comm) {
+  SizeType num_elements = c_tile.ld() * c_tile.size().cols();
+  void const* buf = c_tile.ptr(TileElementIndex(0, 0));
+  MPI_Datatype dtype = dlaf::comm::mpi_datatype<ScalarType>::type;
```

Comment:
I was considering using it but I wasn't sure if there are any allocations being made internally. In my version, I am technically sending more data than I need to for tiles on the edge whose size is different than the size of the tiles specified when constructing the matrix. I do that instead of allocating a contiguous buffer (staging area) that offloads only the useful part of the edge tile and sends that. Is dlaf::comm::Message going to do what I do or allocate a staging area?

Reply:
The basic ideas behind this are:
- The `Tile` manages the underlying allocated memory
- The `Data` concept allow the generalization of a memory layout: it does not allocate any additional memory, it just refers to the data and gives you access to the underlying layout properties in a "generic" way (in terms of number of blocks, number of elements per block, stride between blocks);
- The `Message` allows you to make a `Data` concept compatible with the MPI library, internally computing what type is the best to send that particular layout via MPI calls.

For reference, you can give a look at the recently merged PR #155.

About edge cases, where `Tile` may have a different size, I want to highlight this code snippet
https://github.com/eth-cscs/DLA-Future/blob/ec93197fa42d99bc3753eea6e0b9bd36b4828ded/include/dlaf/tile.h#L185-L190

This is the code that starting from a `Tile`, it creates a `Data` concept for it. As you can see it uses tile sizes to describe the layout of the memory. So, I would say that it should create a `Message` from it with just needed data.

I'm willing to get a feedback from you on this 😉